mlir/lib/Dialect/Rock/Transforms/RockPipeline.cpp

```diff
@@ -0,0 +1,438 @@
+constexpr DependencyType RAR{MemoryAccessType::READ, MemoryAccessType::READ};
+
+using ScheduleType = std::vector<std::pair<Operation *, unsigned>>;
+using DagType =
+    DenseMap<rock::StageOp,
+             DenseMap<rock::StageOp,
+                      DenseSet<std::pair<rock::GpuAllocOp, DependencyType>>>>;
+
+namespace llvm {
+template <> struct DenseMapInfo<MemoryAccessType> {
+  using StorageInfo = ::llvm::DenseMapInfo<uint32_t>;
+
+  static inline MemoryAccessType getEmptyKey() {
+    return static_cast<MemoryAccessType>(StorageInfo::getEmptyKey());
+  }
+
+  static inline MemoryAccessType getTombstoneKey() {
+    return static_cast<MemoryAccessType>(StorageInfo::getTombstoneKey());
+  }
+
+  static unsigned getHashValue(const MemoryAccessType &val) {
+    return StorageInfo::getHashValue(static_cast<uint32_t>(val));
+  }
+
+  static bool isEqual(const MemoryAccessType &lhs,
+                      const MemoryAccessType &rhs) {
+    return lhs == rhs;
+  }
+};
+
+} // namespace llvm
+
+namespace {
+
+// Simple rewrite pass to remove the stages
+struct RemoveStagesRewritePattern : public OpRewritePattern<rock::StageOp> {
+  using OpRewritePattern<rock::StageOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(rock::StageOp op,
+                                PatternRewriter &rw) const override {
+    Block *sourceBlock = &op.getRegion().front();
+    rw.eraseOp(sourceBlock->getTerminator());
+    rw.inlineBlockBefore(sourceBlock, op);
+    rw.eraseOp(op);
+    return failure();
+  }
+};
+
+// Given an operation and its operand, find out what kind of access (if any)
+// the operation does on the operand
+MemoryAccessType getOperandAccessType(Operation *op, Value operand) {
+  if (hasEffect<MemoryEffects::Read>(op, operand)) {
+    return MemoryAccessType::READ;
+  } else if (hasEffect<MemoryEffects::Write>(op, operand)) {
+    return MemoryAccessType::WRITE;
+  } else {
+    return MemoryAccessType::UNKNOWN;
+  }
+}
+
+// Create a dependency graph of the given set of stages. The
+// idea is to represent the dependencies through a DAG with
+// the set of shared resources on the the edges
+DagType createDependencyGraph(ArrayRef<rock::StageOp> stages,
+                              const DenseSet<rock::GpuAllocOp> &allocs) {
+  // Mapping resource->[stages using the given resource]
+  DenseMap<rock::StageOp, DenseMap<rock::GpuAllocOp, MemoryAccessType>>
+      resourceMap;
+
+  // Mapping stages->[resources used by the given stage]
+  DenseMap<rock::GpuAllocOp, DenseMap<rock::StageOp, MemoryAccessType>>
+      resourceMapR;
+
+  // For each of the stages walk through the resources they are using. For now
+  // the only type of resource is memory
+  for (auto stage : stages) {
+    stage.walk([&](Operation *op) {
+      for (Value operand : op->getOperands()) {
+        MemoryAccessType accessType = getOperandAccessType(op, operand);
+        auto maybeAlloc = rock::findAlloc(operand);
```

Comment:
This can potentially fail as well ? what happens if so...

Reply:
Well I check if `succeeded(maybeAlloc)` to add the dependency. So if I was not able to find the allocation, I am not adding the dependency.